src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2991,27 +2991,62 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       fun.body match {
         // we can compare arguments and parameters by name because there cannot be a binder between
         // the function's valdefs and the Apply's arguments
-        case Apply(meth, args) if (vparams corresponds args) { case (p, Ident(name)) => p.name == name case _ => false } =>
-          // We're looking for a method (as indicated by FUNmode in the silent typed below),
-          // so let's make sure our expected type is a MethodType
-          val methArgs = NoSymbol.newSyntheticValueParams(argProtos map { case NoType => WildcardType case tp => tp })
-
-          silent(_.typed(meth, mode.forFunMode, MethodType(methArgs, resProto))).fold(EmptyTree: Tree){ methTyped =>
-            if (context.undetparams.isEmpty) {
-              val numVparams = vparams.length
+        // If all vparams are constrained by the method application, see if we can derive expected types for them.
+        // Note that not all method arguments need be references to a function param.
+        case Apply(meth, args) =>
+          // Map param with missing param type to the argument it's passed as in the eta-expanded method application
+          // This list specifies a way to compute the expected parameter type for each of our function's arguments in order.
+          // Either we already know it, and then we have a Type, or we don't, and then it's an index `idx` into
+          // the arguments passed to `meth`, so we can derive it from its MethodType
+          // (based on where the function's parameter is applied to `meth`)
+          val formalsFromApply =
+            vparams.map { vd =>
+              if (!vd.tpt.isEmpty) Right(vd.tpt.tpe)
+              else Left(args.indexWhere {
+                case Ident(name) => name == vd.name
+                case _           => false // TODO: i think we need to deal with widening conversions too??
```

Comment:
I don't understand this comment..

Reply:
Not sure anymore (I wrote that a long time ago), but I think there are certain cases where an `Apply` might wrap the `Ident` in some other tree after `adapt`.

Reply:
Ah OK. In any case it's safe to keep the current version - trying to idenitfy conversions might lead to false positives, and therefore bugs. So currently we (possibly) err on the safe side.

Reply:
probably https://github.com/scala/bug/issues/9738?